### DISCUSSION TITLE
Stable in Docker container

### DISCUSSION THREAD

**Discussion by MrTumnis** [winapps-org/winapps#92] [General]
I do not know why but when I setup Winapps inside a Docker container running a Windows 11 VM, the apps are much more stable. I was getting a lot of screen tearing and glitching when opening and closing apps when compared to using virsh. There were even some apps that would show pop-ups of "you do not have permission" or something along those lines. It was unusable. But now Winapps works fantastic.

**Comment by LDprg** [comment]
This would be worth some testing. Since we could also use podman for userspace only implementation.

@AkechiShiro @oskardotglobal What do you think?

**Reply by AkechiShiro** [reply]
Yes I totally agree, we need to work on this

**Reply by oskardotglobal** [reply]
Would be ideal to set up the required changes in the VM automatically. We could also easily install e.g. an SSH server on the guest to easily install software and do configuration

**Reply by LDprg** [reply]
I would use [this](https://hub.docker.com/r/dockurr/windows) image since it allows different windows versions (even tiny ones) to be run. And allows easier modification of settings from a software side (there seems even docs about how to do it).

It claims to have hw acceleratorion, this however does need testing. And we should also inspect if we can get gpu/hw passthrough to work (for gaming).

**Reply by oskardotglobal** [reply]
Seems perfect as we can just run a shell script after installation to install openssh & directly configure the VM with rdp from there

**Reply by oskardotglobal** [reply]
It also seems that RDP is enabled out of the box

**Comment by LDprg** [comment]
@MrTumnis which image are you using?

**Reply by MrTumnis** [reply]
I am using the dockurr/windows image you pointed out

**Comment by oskardotglobal** [comment]
@MrTumnis If I may ask you some questions:
Could you describe your experience setting it up? Was it easy? Any caveats? How does performance look like?
How much disk space does it need?

**Reply by MrTumnis** [reply]
It is relatively easy setting it up using the image provided by dockurr/windows. Personally I used the .yaml file so that I could make quick and memorable changes when needed. You can even edit the xml file to use the current user instead of the default "docker" user. I have also found it to start the VM much faster than virsch

**Reply by MrTumnis** [reply]
Also, I am using 2 cpu, 4 ram, and 200gb of space. Seems to run very well. The one caveat may be that you would need to manually add a usb to the file in order to redirect it. I have tried several methods in the yaml and through bash scripts to try to automatically redirect without needing to stop the VM first, but so far no luck. I am novice when it comes to docker, so I am sure there are ways this can be done

**Reply by AkechiShiro** [reply]
Add a usb to the file in order to redirect it ? can you elaborate more I don't understand sorry.

**Reply by MrTumnis** [reply]
I should have clarified. As explained in dockur/windows instructions, to pass a usb into the container, you would need to add a line to the yaml file like so:

environment:
  ARGUMENTS: "-device usb host,vendorid=0x1234,productid=0x1234"
 devices: 
    - /dev/bus/usb

So instead of passing the usb to the VM via the redirect like you would in virsch, you would need to specify the usb in the yaml, and then stop/start the container. But I do think there is a workaround, I just haven't found a viable solution

**Reply by LDprg** [reply]
virsh automatically handles usb devices. With the usage of docker-qemu we need to manually pass the device through. The nice thing about this is that this also works for gpu passthrough and other pcie devices.

So we could basically add a cli/ui for adding devices (usb, vifo/gpu/pcie) and automate it easily.

Another thing is that we probably should set the container to privileged. This throws away some extra security (extra container for qemu), but allows us to just pass through options for qemu and let the docker image do its thing on everything else automatically. We also get around some rare platform specific problems, expecially regarding networking.

We also won't need to care about windows images since this is part of the docker image. We just should create a docker volume for windows image caching and data storage. And add an option to use customised images.

We also get noVnc as backup remote for free. 

Another feature that would fit nicely would a special low latency mode by leveraging looking glass to stream the full screen for gaming. Just so you can add games to you desktop and launch them (probably via ssh) automatically on the windows machine (the missing app streaming wouldn't matter since most games are full screen).

I would even propose to drop other vm/virsh/quickemu implementations for the rewrite completely. Those docker images are just preconfigured kvm and codec using qemu instances. So there shouldn't be any performance downsides.

For the existing winapps scripts we could just write a new tutorial. It would simplify installation a lot and would allow more customisation.

On my machine it works great (no usb, pcie passthrough yet) with the old scripts. We also get rid of the quickemu dependency management.

Sorry for the wall of text, but I just want to summarise my results of my investigations. ;-)

**Reply by AkechiShiro** [reply]
Thanks @LDprg and @MrTumnis for your explanations and details.
